kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceBatchAction.java

```diff
@@ -0,0 +1,434 @@
+    }
+
+    if (request.getEntries().size() > BATCH_MAXIMUM_ENTRIES) {
+      throw Errors.invalidPayloadException(
+          "More than " + BATCH_MAXIMUM_ENTRIES + " entries in batch.");
+    }
+
+    List<ProduceBatchResultEntry> successes = new LinkedList<>();
+    List<ProduceBatchErrorEntry> failures = new LinkedList<>();
+
+    ProduceController controller = produceControllerProvider.get();
+    CompletableFuture.allOf(
+            request.getEntries().stream()
+                .map(
+                    entry ->
+                        produce(
+                                clusterId,
+                                topicName,
+                                entry,
+                                controller,
+                                producerMetricsProvider.get())
+                            .whenComplete(
+                                (result, error) -> {
+                                  if (error != null) {
+                                    failures.add(
+                                        toErrorEntryForException(entry.getId().textValue(), error));
+                                  } else {
+                                    successes.add(result);
+                                  }
+                                }))
+                .toArray(CompletableFuture[]::new))
+        .whenCompleteAsync(
+            (result, error) -> {
+              asyncResponse.resume(
+                  Response.status(207, "Multi-Status")
+                      .entity(
+                          ProduceBatchResponse.builder()
+                              .setSuccesses(successes)
+                              .setFailures(failures)
+                              .build())
+                      .build());
+            },
+            executorService);
+  }
+
+  private CompletableFuture<ProduceBatchResultEntry> produce(
+      String clusterId,
+      String topicName,
+      ProduceBatchRequestEntry request,
+      ProduceController controller,
+      ProducerMetrics metrics) {
+    final long requestStartNs = System.nanoTime();
+
+    try {
+      try {
+        produceRateLimiters.rateLimit(clusterId, request.getOriginalSize());
+      } catch (RateLimitExceededException e) {
+        recordRateLimitedMetrics(metrics);
+        // KREST-4356 Use our own CompletionException that will avoid the costly stack trace fill.
+        throw new StacklessCompletionException(e);
+      }
+
+      // Request metrics are recorded before we check the validity of the message body, but after
+      // rate limiting, as these metrics are used for billing.
+      recordRequestMetrics(metrics, request.getOriginalSize());
+
+      request
+          .getPartitionId()
+          .ifPresent(
+              (partitionId) -> {
+                if (partitionId < 0) {
+                  throw Errors.partitionNotFoundException();
+                }
+              });
+
+      Optional<RegisteredSchema> keySchema =
```

Comment:
This might be out of scope, I'm thinking in the future, once we integrate schema registry, should we require that all records in a batch adhere to one schema?